### ISSUE TITLE
Accessibility color themes

### ISSUE THREAD

**Issue by HeyMyian** [vkbo/novelWriter#2407]
While I was tweaking my contribution of the Primer themes, I concerned myself with the question of color contrast accessibility standards. Researching this, I found that the WCAG 2.x standard, on which the most popular online contrast checkers are based, is outdated. WCAG 2.x is known to judge color contrasts with a ratio, e.g. `4.5:1` or `12:1`. Apparently, for specific colors/brightnesses, this system produces "passing" contrast ratios that are however very hard to read by the human eye.

The new system is called APCA and the specification is a little dense. As I understand(?) [the gist of it](https://git.apcacontrast.com/documentation/APCAeasyIntro#use-case-ranges), the minimum "readable" contrast is Lc&nbsp;60, and minimum for body text is Lc&nbsp;75.

(**Edit:** ACPA is not part of any working standards, in fact is has been removed from the WCAG 3 draft. There appear to be quite some controversies. I was too fast to champion it as new standard.)

An APCA contrast checker can be found e.g. here: https://contrast.tools/?tab=apca
There is also a WCAG/APCA switch to compare the two standards. For example, the default dark theme colors `base` `#373737` and `default` `#cccccc` easily pass WCAG but technically fail in APCA.

Now, I don't know if or to which extent the novelWriter themes should be concerned with this because:

- There are many different themes the user can choose from.
- The user can set their own preferred font and font size, which effect the contrast rating.
- When foreground and background color meet the required minimum contrast, the contrast between the different *syntax highlighting* colors in turn might not be readily perceivable.

_Originally posted by @HeyMyian in https://github.com/vkbo/novelWriter/discussions/2381#discussioncomment-13471730_

**Comment by vkbo** [comment]
Did any of your themes fulfil the contrast requirements?

If so, maybe they should be labelled as such.

**Comment by HeyMyian** [comment]
I just double-checked the themes I submitted, and generally: none fulfil the requirements.

It's kind of impossible to tell since we won't know the user's font size, font weight and font family, which all play an important part in the ACPA specification. Visually bolder and larger fonts with more pixel coverage per letter will receive a better rating.

I tried checking for WCAG because they give a very clear yes/no answer, but then I was reminded that [this contrast](https://contrast.tools?tab=apca&text=878a91&background=1c1f28) "passes" (in quotation marks) WCAG AA but completely fails in the new ACPA and which I agree is not an accessible contrast. It's a `faded` text in a dark theme.

Here is an ACPA contrast checker that will display passing font sizes and weights (generalized, because font family matters too): https://apcacontrast.com/

There seems to be always some parts of the text in the editor that will fail the spec - because they are *supposed* to stand back, visually. Like comments or maybe shortcodes... Or sacrificed their contrast against the background in order to improve contrast against other text elements.

I'm not sure it's possible to make a color-based syntax highlighting theme accessible according to these specifications (actually neither WCAG *nor* ACPA). Maybe if one were to reduce the number of used colors to three (yellow, cyan, and green, they are excellent neon colors on `#000000`), or if we forego accessible contrasts for certain less important text elements.

That's kind of why I formulated my original text here as question. novelWriter is not a website; the users who need it will use a good font and a larger font size anyway. Maybe the best that can be done is to provide a dark theme with neon colors - or use a different highlighting method, like different fonts - or to provide an option to appropriate text background (like in `texthighlight`) as a general highlighting method instead of text color.

**Comment by vkbo** [comment]
The question then is whether we should make two dedicated high contrast themes for the purpose. Since the font is set independently, I guess picking a reasonable font with a sensible size to pass the requirements is enough.

**Comment by HeyMyian** [comment]
(ACPA is not part of any working standards, in fact is has been removed from the WCAG 3 draft. There appear to be quite some controversies. I was too fast to champion it as new standard.)

> The question then is whether we should make two dedicated high contrast themes for the purpose. Since the font is set independently, I guess picking a reasonable font with a sensible size to pass the requirements is enough.

I'm also inclined to think that it is not necessary. There are now a couple of more themes that provide a better contrast than the default dark theme. Light themes are a different beast since it's so hard to assemble high-contrasting colors that are actually colorful.

What was your idea for labeling these themes?

**Comment by vkbo** [comment]
> What was your idea for labeling these themes?

Just adding "High Contrast" in parenthesis to the name for a couple of themes specifically designed for the purpose.

**Comment by vkbo** [comment]
I think with the new high contrast themes you've added, this can be closed as completed?

**Comment by HeyMyian** [comment]
Yes, I'm inclined to think so, too.

**Comment by vkbo** [comment]
Great, I'll close it then. I'll link it to the PR too so it shows up there when I write the changelog for 2.8.
